src/pubtools/_pulp/tasks/common.py

```diff
@@ -119,9 +120,15 @@ def publish_with_cache_flush(
         # publish the repos found
         publish_fs = self.publish(repos)
 
-        # wait for the publish to complete before
-        # flushing caches.
-        f_sequence(publish_fs).result()
+        if publish_fs and all(isinstance(f, Future) for f in publish_fs):
```

Comment:
This fix feels quite ugly and risky.  Risky because it can hide type errors, and because skipping await of publish in cases where we should not is potentially very severe (e.g. exodus-gw commit will be attempted before pulp tasks have finished).

A very similar issue already came up for the `flush_cdn` method (which has since been deleted).  I think we should fix it the same way it was fixed there, in https://github.com/release-engineering/pubtools-pulp/pull/266 .

Essentially you do:

```diff
--- a/src/pubtools/_pulp/tasks/common.py
+++ b/src/pubtools/_pulp/tasks/common.py
@@ -70,7 +70,7 @@ class Publisher(UdCache):
             action="store_true",
         )
 
-    @step("Publish")
+    @step("Publish", skipped_value=[])
     def publish(self, repos):
         out = []
```

This tells the `@step` decorator "if the step is skipped, the returned value should be an empty list".  An empty list represents that zero publishes occurred.  All the existing code here should just work, as it's entirely valid to do `f_sequence(publish_fs).result()` on an empty list.

Reply:
Thanks, I'll give that a try

Reply:
Looks good!